src/main/scala/tech/sourced/gemini/Gemini.scala

```diff
@@ -16,10 +16,23 @@ import scala.io.Source
 
 class Gemini(session: SparkSession) {
 
-  def hash(reposPath: String): DataFrame = {
+  def hash(reposPath: String, limit: Int = 0): DataFrame = {
     val engine = Engine(session, reposPath)
 
-    val headRefs = engine.getRepositories
+    // engine.getRepositories.limit(n)...getFiles - doesn't work in engine now
+    // use normal select without limit and filter + union with filter
+    if (limit <= 0) {
+       hashForRepos(engine.getRepositories)
+    } else {
+      engine.getRepositories.limit(limit).select("id").collect().map(r => {
+        val id = r(0)
+        hashForRepos(engine.getRepositories.filter(s"id = '$id'"))
+      }).reduce((df1, df2) => df1.union(df2))
```

Comment:
1. It would be more idiomatic to keep chaining on separate lines each i.e

```scala
engine....collect()
 .map { r =>
   ...
 }
 .reduce()
```

2. instead of using strings for expressions, let's more safe [SQL expression](https://jaceklaskowski.gitbooks.io/mastering-apache-spark/spark-sql-Dataset.html) syntax with [short-hand for a named columns](https://spark.apache.org/docs/2.2.0/api/scala/index.html#org.apache.spark.sql.Column)

```scala
import session.implicits._ // in Gemini, or `import spark.implicits._` in a spark-shell

.filter($"type" === id)
```

3. As a workaround for https://github.com/src-d/engine/issues/267 instead of calling `hashForRepos` in the loop, let's call it once and pass in the collected IDs using [isin()](https://spark.apache.org/docs/2.2.0/api/scala/index.html#org.apache.spark.sql.Column@isin(list:Any*):org.apache.spark.sql.Column)

```scala
.where($"id".isin(ids:_*))
```

Good exercise would be - to observe the difference implied by such change though running both versions on `spark-shell`, watching resulting jobs in WebUI and printing the [logical and physical execution plans](https://jaceklaskowski.gitbooks.io/mastering-apache-spark/spark-sql-SparkPlan.html) for dataframes using `.explain()`